pkg/crc/api/api_http_test.go

```diff
@@ -1,314 +1,507 @@
+	failRequest bool
+	response    response
 }
 
-func TestDelete(t *testing.T) {
-	client := newTestClient()
-	defer client.Close()
-	deleteResult, err := client.Delete()
-	assert.NoError(t, err)
-	assert.Equal(
-		t,
-		apiClient.Result{
-			Success: true,
-			Error:   "",
-		},
-		deleteResult,
-	)
+func get(resource string) request {
+	return request{
+		httpMethod: http.MethodGet,
+		resource:   resource,
+	}
 }
 
-func TestConfigGet(t *testing.T) {
-	client := newTestClient()
-	defer client.Close()
-	configGetResult, err := client.GetConfig([]string{"cpus"})
-	assert.NoError(t, err)
-	assert.Equal(
-		t,
-		apiClient.GetConfigResult{
-			Success: true,
-			Error:   "",
-			Configs: map[string]interface{}{
-				"cpus": float64(4),
-			},
-		},
-		configGetResult,
-	)
+func post(resource string) request {
+	return request{
+		httpMethod: http.MethodPost,
+		resource:   resource,
+	}
 }
 
-func TestConfigSet(t *testing.T) {
-	client := newTestClient()
-	defer client.Close()
-	configSetResult, err := client.SetConfig(apiClient.SetConfigRequest{
-		Properties: map[string]interface{}{
-			"cpus": float64(5),
-		},
-	})
-	assert.NoError(t, err)
-	assert.Equal(
-		t,
-		apiClient.SetOrUnsetConfigResult{
-			Success:    true,
-			Error:      "",
-			Properties: []string{"cpus"},
-		},
-		configSetResult,
-	)
-	// Get the cpus config again after setting it to make sure it is set
-	// properly.
-	configGetAfterSetResult, err := client.GetConfig([]string{"cpus"})
-	assert.NoError(t, err)
-	assert.Equal(
-		t,
-		apiClient.GetConfigResult{
-			Success: true,
-			Error:   "",
-			Configs: map[string]interface{}{
-				"cpus": float64(5),
-			},
-		},
-		configGetAfterSetResult,
-	)
+func delete(resource string) request {
+	return request{
+		httpMethod: http.MethodDelete,
+		resource:   resource,
+	}
 }
 
-func TestConfigUnset(t *testing.T) {
-	client := newTestClient()
-	defer client.Close()
-	configUnsetResult, err := client.UnsetConfig([]string{"cpus"})
-	assert.NoError(t, err)
-	assert.Equal(
-		t,
-		apiClient.SetOrUnsetConfigResult{
-			Success:    true,
-			Error:      "",
-			Properties: []string{"cpus"},
-		},
-		configUnsetResult,
-	)
+func (req request) String() string {
+	return fmt.Sprintf("%s /%s HTTP/1.1", req.httpMethod, req.resource)
 }
 
-func TestConfigGetAll(t *testing.T) {
-	client := newTestClient()
-	defer client.Close()
-	allConfigGetResult, err := client.GetConfig(nil)
-	assert.NoError(t, err)
-	configs := make(map[string]interface{})
-	for k, v := range client.config.AllConfigs() {
-		// This is required because of https://pkg.go.dev/encoding/json#Unmarshal
-		// Unmarshal stores float64 for JSON numbers in case of interface.
-		switch v := v.Value.(type) {
-		case int:
-			configs[k] = float64(v)
-		default:
-			configs[k] = v
-		}
+func (req request) withBody(data string) request {
+	req.data = data
+	return req
+}
+
+func jSon(data string) response {
+	return response{
+		statusCode: 200,
+		protoMajor: 1,
+		protoMinor: 1,
```

Comment:
@cfergeau what does this protoMajor/protoMinor for? It is the api version scheme?

Reply:
It is http protocol version, the `HTTP/1.1` at the end of each requests. Might change to 2.0 some day. 
Reminds me at least one thing which is missing, it is a check for the `Content-Type` header.

Reply:
@cfergeau that we can improve later also.